cfront_3/man/man3/Args.3
.\" ident	@(#)Args:man/Args.3	3.2
.\"
.\" C++ Standard Components, Release 3.0.
.\"
.\" Copyright (c) 1991, 1992 AT&T and UNIX System Laboratories, Inc.
.\" Copyright (c) 1988, 1989, 1990 AT&T.  All Rights Reserved.
.\"
.\" THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T and UNIX System
.\" Laboratories, Inc.  The copyright notice above does not evidence
.\" any actual or intended publication of such source code.
.\" 
.TH \f3Args\fP \f33C++\fP  " "
.SH NAME
Args \- UNIX command line arguments
.SH SYNOPSIS of Args.h
.Bf
#include <Objection.h>   // see \f3Objection(3C++)\fP

class Args {
public:

//  Constructor

    enum { intermix = ..., plusoptions = ... };
    Args(int argc, const char*const* argv,
        const char* optstr, int bits = 0,
        const char*const* keyword_options = 0);

//  Test options and get values

    int isset(char opt) const;
    int isset(const char* key) const;
    char flag(char opt) const;
    char flag(const char* key) const;
    const char* value(char opt) const;
    const char* value(const char* key) const;
    const char* subvalue(char opt,
        const char* name) const;
    const char* subvalue(const char* key,
        const char* name) const;
    const char* arg(int i) const;
    const char* progname() const;
    int nargs() const;
    int error() const;

//  Objections

    static Objection unexpected_opt;
    static Objection missing_opt;
    static Objection missing_val;
    static Objection opt_as_val;
};

class Opt {
public:
    char flag() const;
    char chr() const;
    const char* keyword() const;
    const char* value() const;
};

class Subopt {
public:
    const char* name() const;
    const char* value() const;
};

class Optsiter {
public:
    Optsiter(const Args& args);
    int next(const Opt*& opt);
    const Args* the_args() const;
};

class Suboptsiter {
public:
    Suboptsiter(const Opt& opt);
    int next(const Subopt*& subopt);
    const Opt* the_opt() const;
};

class Argsiter {
public:
    Argsiter(const Args& args);
    int next(const char*& arg);
    const Args* the_args() const;
};
.fi
.ft R
.SH DESCRIPTION
.PP
A UNIX command is invoked with a series of 
.I options 
and non-option 
.IR arguments .
An option is a letter (or keyword; see below)
preceded by a ``\f4-\fP'' (and optionally by a ``\f4+\fP''; see below).
The ``\f4-\fP'' (and optionally ``\f4+\fP'') character
is called the option's
.IR flag .
An option may be followed by a character string, called its
.IR value .
White space between an option and its value is optional.
.PP
A collection of options, none of which have values, 
may be specified in an
option
.IR block , 
for example,
``\f4-abc\fP''.
Alternatively, they may be specified individually; for example,
``\f4-a -b -c\fP''.
The two forms are equivalent.
(Of course, if ``\f4+\fP'' is specified as an option flag, 
then ``\f4+\fP'' and ``\f4-\fP'' 
options must be specified separately.)
.PP
Some option values are further broken up into 
.I suboptions 
(see below).
.PP
The leftmost command line argument that is neither an option, option block,
nor option value is an 
.IR argument .
All command line arguments to the right of the leftmost argument are
also arguments, regardless
of whether they start with a flag.
However, if intermixing
of options and arguments is desired, this can be optionally specified.
In either case, the special option ``\f4--\f1'' can be used to indicate that all
command line arguments to the right are to be treated as arguments,
regardless of whether they start with a flag.
.PP
The \f4Args\fP class scans the command line and finds all
of the options, values, suboptions, and arguments.
These are saved in the \f4Args\fP object and can be subsequently
examined (but not modified).
.PP
.SS Keyword options
Options consisting of more than a single character are called
.I keyword options.
For example, on some machines, in
.ft CW

    cc -dryrun -c foo.c

.ft P
``dryrun'' is a keyword option.
Specifying a string as a keyword option to Args causes
any command line argument which matches that string (prefixed by a flag)
to be treated as that keyword option, 
even if it consists of a sequence of legal
single-character options.
.SS Suboptions
The value of an option specified as having suboptions
is assumed to be a series
of one or more strings separated by ``\f4,\fP''s with
no intervening spaces.
Each suboption can be either a simple string 
.I name
of the suboption),
or a name followed by an ``\f4=\fP'' followed
by another string (the
.I value
of the suboption).
.PP
For example, suppose a program was invoked as follows:
.ft CW

    foo \-O alpha=tom,beta,gamma=harry

.ft P
Assuming that option ``\f4O\fP'' was specified as having suboptions,
the first suboption has name ``\f4alpha\fP'' and value ``\f4tom\fP'',
the second suboption has name ``\f4beta\fP'' and value ``'' (the null
string), and the third suboption has name ``\f4gamma\fP'' and value
``\f4harry\fP''.
.SH " "
.SH "Args"
.SH " "
.SS "Constructors"
.IP "\f4enum { intermix = ..., plusoptions = ... };\f1"
.hS
.IP "\f4Args(int argc, const char*const* argv,\f1"
.IC "\f4    const char* optstr, int bits = 0,\f1"
.IC "\f4    const char*const* keyword_options = 0);\f1"
The first two arguments are the \f4argc\fP and
\f4argv\fP arguments with which \f4main\fP is invoked.
Each non-keyword option acceptable to the program must be contained in
the string pointed to by 
\f4optstr\fP.
If the option is to have a value, the letter must be followed 
by a ``\f4:\fP''.
If the option is to have suboptions, the letter must be followed
by a ``\f4;\fP''.
If \f4bits\ |\ intermix\f1 is true, then intermixing of options and arguments is allowed.
If \f4bits\ |\ plusoptions\f1 is true, then ``\f4+\f1'' is treated as an option flag.

The optional \f4keyword_options\fP is a pointer to a NULL-terminated
list of allowed
keyword options.  
If a keyword in this list is to have a value, its
final character should be a ``\f4:\fP''.
If a keyword in this list is to have suboptions, its
final character should be a ``\f4;\fP''.
(The final ``\f4:\fP'' or ``\f4;\fP'' is not considered part of the keyword.)
.SS "Test options and get values"
.IP "\f4int isset(char opt)const;\f1"
.hS
.IP "\f4int isset(const char* key)const;\f1"
Returns nonzero if the (keyword) option appeared on the command line.
.IP "\f4char flag(char opt)const;\f1"
.hS
.IP "\f4char flag(const char* key)const;\f1"
Returns the flag with which the rightmost occurrence of the
(keyword) option on the command line was set.
If the option was not set, the character ``\f4\e0\fP'' is returned.
.IP "\f4const char* value(char opt)const;\f1"
.hS
.IP "\f4const char* value(const char* key)const;\f1"
Returns a pointer to the value specified with the rightmost occurrence of 
the (keyword) option
on the command line.
If the option was not set or no value was specified,
the 0 pointer is returned.
.IP "\f4const char* subvalue(char opt, const char* name)const;\f1"
.hS
.IP "\f4const char* subvalue(const char* key,\f1"
.IC "\f4    const char* name)const;\f1
.br
Returns a pointer to the string that was 
specified as the value of the rightmost occurrence of the suboption with 
the given \f4name\fP for the rightmost occurrence of
(keyword) option
on the command line.
If the \f4name\fP was given but no value provided,
the return value will be a pointer to the null string.
If the option was not specified, or the \f4name\fP was not
specified in the rightmost occurrence of this option, 
then the 0 pointer is returned.
.IP "\f4const char* arg(int i)const;\f1"
Returns the i'th (non-option) argument, or 0 if there was no such argument. 
The arguments are numbered starting at zero.
Note that \f4argv[0]\fP, conventionally the program name,
is not accessible via this function.
(See \f4progname\fP.)
.IP "\f4const char* progname()const;\f1"
Returns a pointer to the program name, \f4argv[0]\fP.
.IP "\f4int nargs()const;\f1"
Returns the number of (non-option) arguments.
Note that \f4argv[0]\fP, conventionally the program name,
is not included in this count.
.IP "\f4int error()const;\f1"
Returns nonzero if any of the Objections
\f4unexpected_opt\f1,
\f4missing_opt\f1, or
\f4missing_val\f1
were raised.
Programs can use this if they wish to exit in response to option errors.
.SS Objections
.IP "\f4static Objection unexpected_opt;\f1"
Raised when an option not specified in the \f4legal_opts\fP
list is encountered in the input.
The default and recovery action is to print a message and ignore the 
unexpected option.
.IP "\f4static Objection missing_opt;\f1"
Raised when "\f4-\f1" appears by itself, without any following option letters.
The default and recovery action is to print a message and ignore the ``\f4-\f1''.
.IP "\f4static Objection missing_val;\f1"
Raised when an option specified in the \f4legal_opts\fP
list as requiring a value (i.e., followed by a ``\f4:\fP'')
is encountered in the input
without a value.
This can occur if the option requiring a value is in an option block,
or is the last field on the command line.

Note that \f3getopt(3C)\fP does not currently enforce the
requirement that options taking values cannot be mixed with options that
do not take values.
However, it does state the requirement and warns that it may be enforced
in a future release.
We enforce it now since failure to do so would conceal important errors.

The default and recovery action is to print a message and ignore the option.
.IP "\f4static Objection opt_as_val;\f1"
Raised when the value for an option begins with a flag character.
This usually means that the value for the option was omitted.
However, in order to be consistent with UNIX handling of this case,
the default and recovery actions are to ignore the (probable) error
and treat the
string beginning with the flag character as the
value of the preceding option.

An application can catch this Objection to handle the case differently.
The argument passed to the user-appointed handler will be the option
whose value appears to be missing.
If the user handler returns, the suspicious value will be
assigned to this option.
.SS Iterators
Options, suboptions, and arguments
are returned in the order in which they appeared on
the command line.
.SH " "
.SH "Opt"
.SH " "
Value returned by Optsiter (see below).
.IP "\f4char flag() const;"
Returns the flag associated with this option.
.IP "\f4char chr() const;"
If this option is a non-keyword option, returns the option, otherwise 
returns ``\f4\e0\fP''.
.IP "\f4const char* keyword() const;"
If this option is a keyword option, returns a pointer to the
keyword, otherwise returns 0.
.IP "\f4const char* value() const;\f1"
Returns a pointer to the value associated with this option, or 0
if no value was specified.
.SH " "
.SH "Subopt"
.SH " "
Value returned by Suboptsiter (see below).
.IP "\f4const char* name() const;\f1"
Returns a pointer to the name of this suboption.
.IP "\f4const char* value() const;\f1"
Returns a pointer to the value of this suboption.
If no value was provided, 
the return value will be a pointer to the null string.
.SH " "
.SH "Optsiter"
.SH " "
.IP "\f4Optsiter(const Args& args);\f1"
Iterates over all the options.
.IP "\f4int next(const Opt*& opt);\f1"
Sets \f4opt\f1 to a pointer to an Opt object
describing the next option, or 0 if there are no more options.
Returns nonzero if there was another option.
.IP "\f4const Args* the_args()const;\f1"
Returns a pointer to the \f4Args\fP object being iterated over.
.SH " "
.SH "Suboptsiter"
.SH " "
.IP "\f4Suboptsiter(const Opt& opt);\f1"
Iterates over all of the suboptions of the given option.
.IP "\f4int next(const Subopt*& subopt);\f1"
Sets \f4subopt\f1 to a pointer to a Subopt object
describing the next suboption, or 0 if there are no more suboptions.
Returns nonzero if there was another suboption.
.IP "\f4const Opt* the_opt()const;\f1"
Returns a pointer to the \f4Opt\fP object being iterated over.
.SH " "
.SH "Argsiter"
.SH " "
.IP "\f4Argsiter(const Args& args);\f1"
Iterates over all the (non-option) arguments.
.IP "\f4int next(const char*& arg);\f1"
Sets \f4arg\f1 to a pointer to the next (non-option) argument.
Returns nonzero if there was another argument.
.IP "\f4const Args* the_args()const;\f1"
Returns a pointer to the \f4Args\fP object being iterated over.
.SH SEE ALSO
.Bf
\f3Objection(3C++)\fP, \f3getopt(3C)\fP
.Be


cfront_3/tools/demangler/osrc/args.c
/*ident	"@(#)cls4:tools/demangler/osrc/args.c	1.3" */

/*******************************************************************************
 
C++ source for the C++ Language System, Release 3.0.  This product
is a new release of the original cfront developed in the computer
science research center of AT&T Bell Laboratories.

Copyright (c) 1993  UNIX System Laboratories, Inc.
Copyright (c) 1991, 1992 AT&T and UNIX System Laboratories, Inc.
Copyright (c) 1984, 1989, 1990 AT&T.  All Rights Reserved.

THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE of AT&T and UNIX System
Laboratories, Inc.  The copyright notice above does not evidence
any actual or intended publication of such source code.

*******************************************************************************/

/*
 * C++ Demangler Source Code
 * @(#)master	1.5
 * 7/27/88 13:54:37
 */
#include <ctype.h>
#if __STDC__ == 1
#	include <stdlib.h>
#else
	extern long strtol();
#endif
#include "elf_dem.h"
#include "String.h"

/* This structure is used to keep
 * track of pointers to argument
 * descriptions in the mangled string.
 * This is needed for N and T encodings
 * to work.
 */
typedef struct {
	char *list[10];
	int pos;
} Place;

static Place here;

/* Strings and flags needed by the argument demangles.  The declarator
 * is built up in ptr.  Type modifiers are held in the flag fields.
 * The type itself is passed in separately.
 */
typedef struct {
	String *ptr;
	int Sign,Uns,Cons,Vol,Ptr,Ref;
} Arg_Remem;

/* initialize Arg_Remem */
static void
mkar(r)
Arg_Remem *r;
{
	r->ptr = mk_String((String *)0);
	r->Sign = r->Uns = r->Cons = r->Vol = r->Ptr = r->Ref = 0;
}

/* free data for Arg_Remem */
static void
delar(r)
Arg_Remem *r;
{
	free_String(&r->ptr);
}

/* This routine formats a single argument
 * on the buffer sptr.
 * c is the type or class name, n is its length.
 */
static void
nsetarg(sptr,r,c,n)
String **sptr;
Arg_Remem *r;
char *c;
{
	r->ptr = nprep_String(c, -1, r->ptr, n);
	if(r->Cons)
		r->ptr = prep_String("const ",6,r->ptr);
	if(r->Vol)
		r->ptr = prep_String("volatile ",9,r->ptr);
	if(r->Uns)
		r->ptr = prep_String("unsigned ",9,r->ptr);
	else if(r->Sign)
		r->ptr = prep_String("signed ",7,r->ptr);
	*sptr = app_String(*sptr,PTR(r->ptr),-1);
	delar(r);
}

/* This routine formats a single argument
 * on the buffer sptr.
 * c is the null terminated type or class name.
 */
static void
setarg(sptr,r,c)
String **sptr;
Arg_Remem *r;
char *c;
{
	nsetarg(sptr,r,c,NAME_MAX);
}


/* Demangle a single function argument.
 * Returns the number of characters processed from c.
 */
int
doarg(sptr,c)
String **sptr;
char *c;
{
	register int i;
	Arg_Remem ar;
	mkar(&ar);

	if(here.pos < 10 && here.pos >= 0)
		here.list[here.pos++] = c;

	for(i=0;c[i];i++) {
		/* Loop until you find a type.
		   Then call setarg and return.
		*/
		switch(c[i]) {

		case 'T':
			{
				Place tmp;
				tmp = here;
				here.pos = c[1] - '1';
				if(here.pos < 0 || here.pos >= tmp.pos-1) {
					delar(&ar);
					return -1;
				}
				doarg(sptr,here.list[here.pos]);
				here = tmp;
				delar(&ar);
				return 2;
			}
		case 'N':
			{
				Place tmp;
				int cycles,pos;
				cycles = c[1] - '0'; pos = c[2] - '1';
				here.pos += cycles - 1;
				tmp = here;
				if(cycles <= 1 || cycles > 9 || pos < 0 || pos >= tmp.pos-1) {
					delar(&ar);
					return -1;
				}
				while(cycles--) {
					here = tmp;
					here.pos = pos;
					doarg(sptr,here.list[here.pos]);
					(*sptr) = app_String(*sptr,",",1);
				}
				*sptr = trunc_String(*sptr, 1);
				here = tmp;
				delar(&ar);
				return 3;
			}

		/* Qualifiers to type names */
		case 'S':
			ar.Sign = 1;
			break;
		case 'U':
			ar.Uns = 1;
			break;
		case 'C':
			ar.Cons = 1;
			break;
		case 'V':
			ar.Vol = 1;
			break;

		/* Pointers, references, and Member Pointers */
		case 'P':
		case 'R':
		case 'M':
			if(ar.Cons) {
				ar.ptr = prep_String(" const",6,ar.ptr);
				ar.Cons = 0;
			}
			if(ar.Vol) {
				ar.ptr = prep_String(" volatile",9,ar.ptr);
				ar.Vol = 0;
			}
			if(c[i] == 'P') {
				ar.ptr = prep_String("*",1,ar.ptr);
				ar.Ptr = 1;  /* hmm, should i set this to zero somewhere, like Cons above? */
			}
			else if(c[i] == 'R') {
				ar.ptr = prep_String("&",1,ar.ptr);
				ar.Ref = 1;  /* same hmm */
			}
			else {
				int cnt = 0;
				char *s;
				ar.ptr = prep_String("::*",3,ar.ptr);
				/* Skip over the 'M' */
				i++;
				cnt = strtol(c+i, &s, 10);
				i = s - c;
				ar.ptr = nprep_String(c+i, -1, ar.ptr, cnt);
				ar.ptr = prep_String(" ",1,ar.ptr);
				i += cnt;
				/* The loop increments i */
				i--;
			}
			break;

		/* Demangle for basic built-in types */
		case 'i':
			setarg(sptr,&ar,"int");
			return i + 1;
		case 'c':
			setarg(sptr,&ar,"char");
			return i + 1;
		case 's':
			setarg(sptr,&ar,"short");
			return i + 1;
		case 'l':
			setarg(sptr,&ar,"long");
			return i + 1;
		case 'f':
			setarg(sptr,&ar,"float");
			return i + 1;
		case 'd':
			setarg(sptr,&ar,"double");
			return i + 1;
		case 'r':
			setarg(sptr,&ar,"long double");
			return i + 1;

		/* Class encodings */
		case '1': case '2': case '3':
		case '4': case '5': case '6':
		case '7': case '8': case '9':
			{
				int cnt = 0;
				char *s;
				cnt = strtol(c+i, &s, 10);
				i = s - c;
				if(strlen(c+i) < cnt) {
					delar(&ar);
					return -1;
				}
				nsetarg(sptr,&ar,c+i,cnt);
				return i+cnt;
			}

		/* Ellipsis and void */
		case 'e':
			if(ar.Cons || ar.Vol || ar.Uns || ar.Sign ||
			   *PTR(ar.ptr) != '\0')
				return -1;
			/* remove previous comma */
			(*sptr) = trunc_String(*sptr,1);
			(*sptr) = app_String(*sptr," ...",4);
			delar(&ar);
			return ++i;
		case 'v':
			if(ar.Vol || ar.Uns || ar.Sign ||
			   *PTR(ar.ptr) == '\0')
				return -1;
			if(ar.Cons) {
				if (ar.Ptr)
					(*sptr) = app_String(*sptr,"const ",6);
				else
					return -1;
			}
			(*sptr) = app_String(*sptr,"void",4);
			(*sptr) = app_String(*sptr,PTR(ar.ptr),-1);
			delar(&ar);
			return ++i;

		/* Arrays */
		case 'A':
			if(*PTR(ar.ptr)) {
				ar.ptr = prep_String("(",1,ar.ptr);
				ar.ptr = app_String(ar.ptr,")",1);
			}
			ar.ptr = app_String(ar.ptr,"[",1);
			{
				int cnt = 0;
				i++;
				while(isdigit(c[i+cnt]))
					cnt++;
				ar.ptr = napp_String(ar.ptr,c+i,-1,cnt);
				i += cnt;
				if(c[i] != '_') {
					delar(&ar);
					return -1;
				}
			}
			ar.ptr = app_String(ar.ptr,"]",1);
			break;

		/* Functions
		 * This will always be called as a pointer
		 * to a function.
		 */
		case 'F':
			ar.ptr = prep_String("(",1,ar.ptr);
			ar.ptr = app_String(ar.ptr,")",1);
			{
				Place tmp;
				tmp = here;
				i++;
				i += doargs(&ar.ptr,c+i);
				if(c[i] != '_') {
					delar(&ar);
					return -1;
				}
				here = tmp;
			}
			break;

		/* Qualified type name
		 */
		case 'Q':
			{
				String *tmp;
				int numquals = 0;
				char *s;
				int j;
				
				tmp = mk_String((String *) 0);
				numquals = strtol(c+i+1, &s, 10);
				
				if (*s != '_') {
					delar(&ar);
					return -1;
				}
				s++;
				
				for (j = 0; j < numquals; j++) {
					int cnt = strtol(s, &s, 10);
					if (cnt <= 0 || strlen(s) < cnt) {
						delar(&ar);
						return -1;
					}
					
					if (j > 0) {
						tmp = app_String(tmp,"::",2);
					}
					tmp = app_String(tmp,s,cnt);
					s += cnt;
				}
				
				i = s - c;
				setarg(sptr,&ar,PTR(tmp));
				free_String(&tmp);
				return i;
			}
			break;
			
		/* Needed when this is called to demangle
		 * an argument of a pointer to a function.
		 */
		case '_':
			delar(&ar);
			return 0;

		default:
			delar(&ar);
			return -1;
		}
	}

	/* Did the argument list terminate properly? */
	{
		int rc = 0;
		if(*PTR(ar.ptr) || ar.Uns || ar.Sign || ar.Cons || ar.Vol)
			rc = -1;
		delar(&ar);
		return rc;
	}
}

/* This function is called to demangle
 * an argument list.
 * Returns the number of characters processed from c.
 */
int
doargs(sptr,c)
String **sptr;
char *c;
{
	int i,n = 0;
	here.pos = 0;

	*sptr = app_String(*sptr,"(",1);
	while(*c && (i = doarg(sptr,c)) > 0) {
		c += i;
		n += i;
		(*sptr) = app_String(*sptr,",",1);
	}

	if(i < 0)
		return -1;

	*sptr = app_String(trunc_String(*sptr, 1),")",1);

	return n;
}
